### DISCUSSION TITLE
Segmentation fault (core dumped)

### DISCUSSION THREAD

**Discussion by Antguz** [davidfrantz/force#128] [Q&A] [answered]
Hi,

I am running FORCE v. 3.7.0 on Ubuntu 20.04 for processing Sentinel imagery using force-level2. I got a weird error message when processing each file. The message says 'Segmentation fault      (core dumped) $EXE $TODO $PRM >> $LOGFILE'. 

The error message seems to change of lines but it has the same root. For instance, after 'processing' the first file the error says

'local:8/0/100%/0.0s /usr/local/bin/force-l2ps_: line 161: 54142 Segmentation fault      (core dumped) $EXE $TODO $PRM >> $LOGFILE'

and then for the second file it says
/usr/local/bin/force-l2ps_: line 161: 53191 Segmentation fault      (core dumped) $EXE $TODO $PRM >> $LOGFILE

I tried to run it using 'sudo force-level2', but the same happened.

Any idea how to solve this?

AG.

**Comment by davidfrantz** [comment]
Hi @Antguz ,

are you running in Docker? If not, please try this first such that I can make sure you run a tested software build.

Cheers,
David

**Reply by Antguz** [reply]
Hi @davidfrantz ,

No, but it was an error in the build. I build it again, It is solved, thanks!

Regards,
AG

**Reply by Antguz** [reply]
Hi @davidfrantz 

Well actually no, the errors are not gone. When I run Landsat 8 imagery there is no problem, but when it starts with the Sentinel imagery the error of 'segmentation fault' starts to appear. Interesting is the there is a Sentinel image without error, but it is out the bounds so it does not export products.

![Screenshot from 2021-10-15 22-15-45](https://user-images.githubusercontent.com/7254767/137571711-e7fa6ff4-1425-4147-a0c5-415f299dff97.png)


What I can do? I am running in Ubuntu 20.04 using the latest FORCE version.

Regards
AG.

**Reply by Antguz** [reply]
I will try using docker

**Reply by davidfrantz** [reply]
Do you have enough memory? ~8GB per process, i.e. 16*8?
Best,
David

**Reply by Antguz** [reply]
Hi @davidfrantz,

Yes, I have enough memory, 128 GB; when I run force-level2 it never goes up the 80%. 

I was trying a Docker image of the 3.7.0 FORCE version and I have the same problem. 

![Screenshot from 2021-10-16 17-27-06](https://user-images.githubusercontent.com/7254767/137814900-575b112c-99ca-4667-95a4-a924339ce70b.png)

Any thought for finding the potential error?

Regards,
AG.

**Reply by davidfrantz** [reply]
Hi @Antguz,
I would start with using ``force-l2ps`` on one of the images that failed, i.e. processing a single image.
If this works, it is likely related to the parallel processing of ``force-level2``.
If it doesn't work, try running this image through the debug Docker (force-debug) and see what this brings up.
Cheers,
David

**Reply by Antguz** [reply]
Hi @davidfrantz,

I tried using `force-l2ps` in docker and local build without parallel processing, and I have the same problem. I have not tried the debug mode in docker yet (sorry I did not find the extension name, force-debug does not work). However, I noticed that the using the same tile number field, the error occurs always in a specific relative orbit number. 

For instance, in the tile T15TWL, all the R069 succeeded and all the R112 failed. The same happened in T15TVL, all the R069 succeeded and all the R112 failed. The R112 raw images seem to be complete, but not R069s. For example, see the raw band-6 below for the same tile. I have downloaded the images twice, so it is not an error of download.

T15TWL_R112
![Screenshot from 2021-10-25 22-15-28](https://user-images.githubusercontent.com/7254767/138803075-aa474560-3940-4ef9-94fd-d8ba48a0d4f8.png)
  
T15TWL_R069
![Screenshot from 2021-10-25 22-15-37](https://user-images.githubusercontent.com/7254767/138803128-fd06a998-4661-408b-9f96-e86b8a6b632c.png)

Any thoughts on how to proceed? thanks in advance.

Regards,

AG.

**Reply by davidfrantz** [reply]
can you send me two image IDs from these two orbits.
And can you please upload your parameterfile here?
I will have a look at soon as I have some time

**Reply by ernstste** [reply]
I'm not sure what causes the segmentation fault.
But the 'missing' data you're seeing in the thumbnail of T15TWL_R069 is simply due to how the orbits overlap with the MGRS tiling grid.

**Reply by Antguz** [reply]
Hi @davidfrantz,

I am sharing 4 four images: two that succeed and two that failed from SEN2A and SEN2B for testing. 

I am following a similar folder structure recommended on the webpage. So feel free to download the whole folder (example) for your convenience and change the 'to_share' path to simulate my settings. The parameter and queue files are in the 'param' folder.

This is the link: https://drive.google.com/drive/folders/164secrulVbgIscKHjZgt8udZ0ogEzeM7?usp=sharing

Thanks in advance for your help!

@ernstste - I understand the orbits and tiling grid situation, but what is odd to me is that the segmentation fault only happens on specific orbits. And it happens in both, local build and docker. That is odd don't you think so?

Regards,
AG.

**Reply by ernstste** [reply]
@Antguz I fully agree. 
'the R112 raw images seem to be complete, but not R069s' sounded to me like you were surprised about the 'missing' data in the images acquired during overpasses in that specific orbit. I probably just misinterpreted your message.

**Comment by davidfrantz** [comment, marked as answer]
Hi @Antguz, I found and fixed the issue.

In a 2D for-loop in the multi-parameter regression-based resolution merging (``RES_MERGE = REGRESSION``), both indices were running to the same N...

```
    for (i=0; i<ny; i++){
    for (j=0; j<ny; j++){
```

should have been:

```
    for (i=0; i<ny; i++){
    for (j=0; j<nx; j++){
```

Strange enough, this method is implemented in FORCE since day 1, but nobody complained yet.
Doesn`t seem a very popular option ;)

Nevertheless, this is fixed now with https://github.com/davidfrantz/force/releases/tag/v3.7.3-fix, and the results look good on my end.
I also noticed that this method was not multi-threaded yet. It is now :)

Thanks for bringing his to my attention.

Cheers,
David

**Reply by Antguz** [reply]
Hi @davidfrantz, 

Thanks so much for your attention on this! We all appreciated your effort on FORCE. I will try to run the new version, if you don't hear from me it is because it works.

Regards,
AG.
